File: src/unifestbestellbot/bot/unknown.py
"""Fallback for messages that no other router/handler claims.

Included LAST in the dispatcher so it never preempts a real flow.
FSM-state handlers (e.g. `category_unrecognized`) match before this
because they bind to a specific state; this catch-all only fires when
no state is active."""

import logging

from aiogram import Router
from aiogram.types import Message
from sqlmodel import Session

from .. import i18n, repo
from ..config import AppConfig
from . import keyboards

log = logging.getLogger(__name__)

router = Router(name="unknown")


@router.message()
async def cmd_unknown(msg: Message, db_session: Session, config: AppConfig) -> None:
    # Ignore channel posts / non-user messages — those never have from_user.
    if msg.from_user is None:
        return

    group = "Unknown"
    reg = repo.registration_for(db_session, msg.from_user.id)
    if reg is not None:
        group = reg.group_name

    text = msg.text or msg.caption or ""
    if msg.edit_date is not None:
        log.warning(
            "⚠️ received edited message %r from chat %s [%s]",
            text, msg.from_user.id, group,
        )
    else:
        log.warning(
            "⚠️ received unrecognized message %r from chat %s [%s]",
            text, msg.from_user.id, group,
        )

    await msg.answer(i18n.UNKNOWN_COMMAND, reply_markup=keyboards.for_user(reg, config))

File: rust/src/bot/unknown.rs
//! Fallback for messages that no other handler claims. Mirrors the
//! Python `bot/unknown.py`.

use sqlx::SqlitePool;

use super::{keyboards, Caller, Outbox};
use crate::config::AppConfig;
use crate::i18n;
use crate::repo;

pub async fn cmd_unknown(
    outbox: &mut Outbox,
    pool: &SqlitePool,
    config: &AppConfig,
    caller: &Caller,
) -> anyhow::Result<()> {
    let reg = repo::registration_for(pool, caller.chat_id).await?;
    outbox.reply(
        i18n::UNKNOWN_COMMAND,
        keyboards::for_user(reg.as_ref(), config),
    );
    Ok(())
}
